File: LQFiles/LQAliPay.m
//
//  LQAliPay.m
//  LQAliSDKDemo
//
//  Created by Artron_LQQ on 2017/11/21.
//  Copyright © 2017年 Artup. All rights reserved.
//

#import "LQAliPay.h"
#import <AlipaySDK/AlipaySDK.h>

@implementation LQAliPay

+ (void)openUrl:(NSURL *)url
        success:(void(^)(NSDictionary *))success
         failed:(void(^)(void))failed {
    if ([url.host isEqualToString:@"safepay"]) {
        
        [[AlipaySDK defaultService] processOrderWithPaymentResult:url standbyCallback:^(NSDictionary *resultDic) {
            [self checkResult:resultDic success:success failed:failed];
        }];
    }
}

+ (void)payOrder:(NSString *)order
       appScheme:(NSString *)scheme
         success:(void(^)(NSDictionary *))success
          failed:(void(^)(void))failed {
    [[AlipaySDK defaultService] payOrder:order fromScheme:scheme callback:^(NSDictionary *resultDic) {
        [self checkResult:resultDic success:success failed:failed];
    }];
}
+ (void)checkResult:(NSDictionary *)result success:(void(^)(NSDictionary *))success failed:(void(^)(void))failed {
    
    if (result) {
        NSString *statusStr = result[@"resultStatus"];
        NSInteger state = statusStr.integerValue;
        if (state == 9000) {
            //支付成功
            NSString *obj = result[@"result"];
            if (obj) {
                
                NSDictionary *dic = [self payResultToDic:obj];
                if (success) {
                    success(dic);
                }
                return;
            }
        }
    }
    
    if (failed) {
        failed();
    }
}
+ (NSDictionary *)payResultToDic:(NSString *)str {
    
    NSString *str1 = [str stringByReplacingOccurrencesOfString:@"\"" withString:@""];
    NSString *str2 = [str1 stringByReplacingOccurrencesOfString:@"{" withString:@""];
    NSString *str3 = [str2 stringByReplacingOccurrencesOfString:@"}" withString:@""];
    
    NSArray *array = [str3 componentsSeparatedByString:@","];
    
    NSMutableDictionary *dic = [NSMutableDictionary dictionary];
    
    for (NSString * tmp in array) {
        NSArray *tmpArray = [tmp componentsSeparatedByString:@":"];
        if (tmpArray.count == 2) {
            NSString *key = tmpArray[0];
            NSString *value = tmpArray[1];
            
            [dic setObject:value forKey:key];
        } else if (tmpArray.count == 3) {
            NSString *key = tmpArray[1];
            NSString *value = tmpArray[2];
            
            [dic setObject:value forKey:key];
        }
    }
    return dic;
}






















@end

File: LQAliSDKDemo/LQAliSDKDemo/LQAliPay.swift
//
//  LQAliPay.swift
//  LQAliSDKDemo
//
//  Created by Artron_LQQ on 2017/11/21.
//  Copyright © 2017年 Artup. All rights reserved.
//

import UIKit

class LQAliPay: NSObject {

    /// 处理支付宝客户端返回的url（在app被杀模式下，通过这个方法获取支付结果）
    ///
    /// - Parameters:
    ///   - url: 支付宝客户端回传的url
    ///   - success: 成功回调
    ///   - failed: 失败回调
    class func open(url: URL, success: ((_ info: [String: String]) -> Void)? = nil, failed: (()->Void)? = nil) {
        
        if let host = url.host {
            if host == "safepay" {
               AlipaySDK.defaultService().processOrder(withPaymentResult: url, standbyCallback: { (info) in
                self.checkResult(info as? [String : Any], success: success, failed: failed)
                })
            }
        }
    }
    
    /// 发起支付
    ///
    /// - Parameters:
    ///   - order: app支付请求参数字符串，主要包含商户的订单信息，key=value形式，以&连接。 由后台返回
    ///   - scheme: 商户程序注册的URL protocol，供支付完成后回调商户程序使用。
    ///   - success: 成功回调
    ///   - failed: 失败回调
    class func pay(order: String, appScheme scheme: String, success: ((_ info: [String: String]) -> Void)? = nil, failed: (()->Void)? = nil) {
        
        AlipaySDK.defaultService().payOrder(order, fromScheme: scheme) { (info) in
            self.checkResult(info as? [String : Any], success: success, failed: failed)
        }
    }
    
    private class func checkResult(_ resultDic: [String: Any]?, success: ((_ info: [String: String]) -> Void)? = nil, failed: (()->Void)? = nil) {
        
        if let result = resultDic {
            if let statusStr = result["resultStatus"] as? String {
                if let status = Int(statusStr) {
                    
                    if status == 9000 {
                        // 支付成功
                        if let obj = result["result"] as? String {
                            
                            let dic = payResultToDic(obj)
                            
                            if let handle = success {
                                handle(dic)
                            }
                            return
                        }
                    }
                }
            }
        }
        
        if let handle = failed {
            handle()
        }
    }
    // 处理支付结果字符串: 看着像个字典, 但是转不成字典, 只能分割字符串后保存到字典内
    private class func payResultToDic(_ str: String) -> [String: String] {
        
        let str1 = str.replacingOccurrences(of: "\"", with: "")
        
        let str2 = str1.replacingOccurrences(of: "{", with: "")
        
        let str3 = str2.replacingOccurrences(of: "}", with: "")
        
        let arr = str3.components(separatedBy: ",")
        
        var dic: [String: String] = [:]
        
        for str5 in arr {
            let results = str5.components(separatedBy: ":")
            
            if results.count == 2 {
                let key = results[0]
                let value = results[1]
                dic[key] = value
            } else if results.count == 3 {
                let key = results[1]
                let value = results[2]
                dic[key] = value
            }
        }
        
        return dic
    }
}
